### PR TITLE
[MXNET-1358]Fit api tutorial

### PR DESCRIPTION
**Pull request by roywei**
## Description ##

1. bring back https://github.com/apache/incubator-mxnet/pull/14462
main contributor @piyushghai 
I have updated the tutorial to the latest fit/estimator api
2. update website API generation( add to API index, fixed imports for automated API doc generation)
3. fixed checkpoint reuqires hybridization

@ThomasDelteil @thomelane @aaronmarkham  

## Preview
* http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-15353/8/tutorials/gluon/fit_api_tutorial.html

* http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-15353/8/api/python/gluon/contrib.html?highlight=estimator#module-mxnet.gluon.contrib.estimator

### CODE REVIEW

File: docs/tutorials/gluon/fit_api_tutorial.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+                          context=ctx)
+
+# Define the handlers, let's say in built Checkpointhandler
+checkpoint_handler = CheckpointHandler(model_dir='./',
+                                       model_prefix='my_model',
+                                       monitor=train_acc,  # Monitors a metric
+                                       save_best=True)  # Save the best model in terms of
+# Let's instantiate another handler which we defined above 
+loss_record_handler = LossRecordHandler()
+# Magic line
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/fit_api_tutorial.md]: Lol ok

File: docs/tutorials/gluon/fit_api_tutorial.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    
+    [Epoch 1] finished in 24.892s: train_accuracy : 0.8826 train_softmaxcrossentropyloss0 : 0.3229 val_accuracy : 0.8474 val_softmaxcrossentropyloss0 : 0.4262 <!--notebook-skip-line-->
+    
+    Train finished using total 50s at epoch 1. train_accuracy : 0.8826 train_softmaxcrossentropyloss0 : 0.3229 val_accuracy : 0.8474 val_softmaxcrossentropyloss0 : 0.4262 <!--notebook-skip-line-->
+
+    Training begin <!--notebook-skip-line-->
+    Epoch 1, loss 0.5741 <!--notebook-skip-line-->
+    Epoch 2, loss 0.3229 <!--notebook-skip-line-->
+
+You can load the saved model, by using ```load_parameters``` API in Gluon. For more details refer to the [Loading model parameters from file tutorial](http://mxnet.incubator.apache.org/versions/master/tutorials/gluon/save_load_params.html#saving-model-parameters-to-file)
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/fit_api_tutorial.md]:
```suggestion
You can load the saved model, by using the `load_parameters` API in Gluon. For more details refer to the [Loading model parameters from file tutorial](save_load_params.html#saving-model-parameters-to-file)
```

File: docs/tutorials/gluon/fit_api_tutorial.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+
+```python
+resnet_18_v1 = vision.resnet18_v1(pretrained=False, classes=10)
+resnet_18_v1.load_parameters('./my_model-best.params', ctx=ctx)
+```
+
+## Summary
+
+In this tutorial, we learnt how to use ```Gluon Fit APIs``` for training a deep learning model and also saw an option to customize it with the use of Event Handlers.
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/fit_api_tutorial.md]:
```suggestion
In this tutorial, you learned how to use Gluon Fit APIs for training a deep learning model and also saw an option to customize it with the use of `Event Handlers`.
```
🇺🇸

File: docs/tutorials/gluon/fit_api_tutorial.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+For more references and advanced usage details can be found in the [documentation](http://mxnet.apache.org/api/python/gluon/gluon.html).
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/fit_api_tutorial.md]:
```suggestion
More references and advanced usage details can be found in the [Gluon API documentation](http://mxnet.apache.org/api/python/gluon/gluon.html).
```

File: docs/tutorials/gluon/fit_api_tutorial.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+## Next Steps 
+
+- To learn more about deep learning with MXNet Gluon, see [Deep Learning - The Straight Dope](https://gluon.mxnet.io)
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/fit_api_tutorial.md]:
```suggestion
- To learn more about deep learning with MXNeT, see [Dive Into Deep Learning](http://gluon.io)
```
D2L > TSD and D2L doesn't have SSL.

File: docs/tutorials/gluon/fit_api_tutorial.md [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+- For more hands on learning about deep learning, check out [Dive into Deep Learning](https://d2l.ai)
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/fit_api_tutorial.md]:
```suggestion

```

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-awaiting-review]

File: docs/tutorials/gluon/fit_api_tutorial.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+
+With the Fit API, you can train a deep learning model with miminal amount of code. Just specify the network, loss function and the data you want to train on. You don't need to worry about the boiler plate code to loop through the dataset in batches(often called as 'training loop'). Advanced users can still do this for bespoke training loops, but most use cases will be covered by the Fit API.
+
+To demonstrate the Fit API, this tutorial will train an Image Classification model using the [ResNet-18](https://arxiv.org/abs/1512.03385) architecture for the neural network. The model will be trained using the [Fashion-MNIST dataset](https://research.zalando.com/welcome/mission/research-projects/fashion-mnist/). 
+
+## Prerequisites
+
+To complete this tutorial, you will need:
+
+- [MXNet](https://mxnet.incubator.apache.org/install/#overview) (The version of MXNet will be >= 1.5.0)
```
[the remark this answers is not in the document]
Comment (roywei) [reply docs/tutorials/gluon/fit_api_tutorial.md]: if build from source it's already 1.5.0 on master, once 1.5.0 release, master will be 1.6.0. Added explaination

File: docs/tutorials/gluon/fit_api_tutorial.md [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+<!--- KIND, either express or implied.  See the License for the -->
+<!--- specific language governing permissions and limitations -->
+<!--- under the License. -->
+
+
+# MXNet Gluon Fit API
+
+In this tutorial, we will see how to use the [Gluon Fit API](https://cwiki.apache.org/confluence/display/MXNET/Gluon+Fit+API+-+Tech+Design) which is the easiest way to train deep learning models using the [Gluon API](http://mxnet.incubator.apache.org/versions/master/gluon/index.html) in Apache MXNet. 
+
+With the Fit API, you can train a deep learning model with miminal amount of code. Just specify the network, loss function and the data you want to train on. You don't need to worry about the boiler plate code to loop through the dataset in batches(often called as 'training loop'). Advanced users can still do this for bespoke training loops, but most use cases will be covered by the Fit API.
```
[the remark this answers is not in the document]
Comment (roywei) [reply docs/tutorials/gluon/fit_api_tutorial.md]: changed to many

File: docs/tutorials/gluon/fit_api_tutorial.md [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+```python
+import mxnet as mx
+from mxnet import gluon
+from mxnet.gluon.model_zoo import vision
+from mxnet.gluon.contrib.estimator import estimator
+from mxnet.gluon.contrib.estimator.event_handler import TrainBegin, TrainEnd, EpochEnd, CheckpointHandler
+
+gpu_count = mx.context.num_gpus()
+ctx = [mx.gpu(i) for i in range(gpu_count)] if gpu_count > 0 else mx.cpu()
+mx.random.seed(7) # Set a fixed seed
```
[the remark this answers is not in the document]
Comment (roywei) [reply docs/tutorials/gluon/fit_api_tutorial.md]: removed it, if a tutorial is been tested against accuracy (e.g. assert acc > 0.8 after 5 epochs, fixed seed may be added for less flakiness). here we are not checking acc, so removed it

File: docs/tutorials/gluon/fit_api_tutorial.md [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+
+Let's load the resnet-18 model architecture from [Gluon Model Zoo](http://mxnet.apache.org/api/python/gluon/model_zoo.html) and initialize it's parameters. The Gluon Model Zoo contains a repository of pre-trained models as well the model architecture definitions. We are using the model architecture from the model zoo in order to train it from scratch.
+
+
+```python
+resnet_18_v1 = vision.resnet18_v1(pretrained=False, classes = 10)
+resnet_18_v1.initialize(init = mx.init.Xavier(), ctx=ctx)
+```
+
+We will be using ```SoftmaxCrossEntropyLoss``` as the loss function since this is a multi-class classification problem. We will be using ```sgd``` (Stochastic Gradient Descent) as the optimizer. You can experiment with a different optimizer as well.
```
[the remark this answers is not in the document]
Comment (roywei) [reply docs/tutorials/gluon/fit_api_tutorial.md]: done

**Comment by roywei** [comment]
@aaronmarkham I have addressed your comments, could you take another look? thanks!
Another problem is the estimator API doc is not showing up on the website, not sure if [this](https://github.com/apache/incubator-mxnet/pull/15353/files#diff-0e3f58347a224fc8184299f9c154cda1R29) will fix it

File: docs/tutorials/gluon/fit_api_tutorial.md [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+
+With the Fit API, you can train a deep learning model with a minimal amount of code. Just specify the network, loss function and the data you want to train on. You don't need to worry about the boiler plate code to loop through the dataset in batches (often called as 'training loop'). Advanced users can train with bespoke training loops, and many of these use cases will be covered by the Fit API.
+
+To demonstrate the Fit API, you will train an image classification model using the [ResNet-18](https://arxiv.org/abs/1512.03385) neural network architecture. The model will be trained using the [Fashion-MNIST dataset](https://research.zalando.com/welcome/mission/research-projects/fashion-mnist/). 
+
+## Prerequisites
+
+To complete this tutorial, you will need:
+
+- [MXNet](https://mxnet.incubator.apache.org/install/#overview) (The version of MXNet will be >= 1.5.0, you can use `pip install mxnet --pre` to get the latest pip package or build from source with master, refer to [MXNet installation](http://mxnet.incubator.apache.org/versions/master/install/index.html?platform=Linux&language=Python&processor=CPU)
```
Comment (abhinavs95) [inline docs/tutorials/gluon/fit_api_tutorial.md]: We can update this now since 1.5 is out?

File: docs/tutorials/gluon/fit_api_tutorial.md [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+gpu_count = mx.context.num_gpus()
+ctx = [mx.gpu(i) for i in range(gpu_count)] if gpu_count > 0 else mx.cpu()
+```
+
+## Dataset
+
+[Fashion-MNIST](https://research.zalando.com/welcome/mission/research-projects/fashion-mnist/) dataset consists of fashion items divided into ten categories: t-shirt/top, trouser, pullover, dress, coat, sandal, shirt, sneaker, bag and ankle boot. 
+
+- It has 60,000 grayscale images of size 28 * 28 for training.  
+- It has 10,000 grayscale images os size 28 * 28 for testing/validation.
```
Comment (abhinavs95) [inline docs/tutorials/gluon/fit_api_tutorial.md]: typo: os->of

**Comment by eric-haibin-lin** [comment]
any update on this ? mxnet 1.5 is released but no doc for the feature.

**Comment by roywei** [comment]
I have addressed commend @aaronmarkham @abhinavs95 could you take another look? Thanks

**Comment by roywei** [comment]
I have verified the documentation is successfully generated, tutorial is passing, nightly test are passing and unit test should be already checked by CI

ping @aaronmarkham for review on the docs. Thanks!

estimator
checkpoint

**Comment by roywei** [comment]
Update website API docs can be found at: http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-15353/10/api/python/gluon/contrib.html

Tutorial: http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-15353/10/tutorials/gluon/fit_api_tutorial.html
